### DISCUSSION TITLE
llm for relation extraction

### DISCUSSION THREAD

**Discussion by AGudden** [explosion/spacy-llm#216] [Q&A] [answered]
I am using the model for enrity and relation extraction. I adjusted the spacy model a little so that I can upload sentences in a csv file. I did some testing and notices that for some sentences I got following error in my terminal: IndexError: tuple index out of range
Closer inspection showed that this is properly caused by sentences that have only 1 entity. I also noticed that the model extraxt relations between entities that are not in my example.json.file.

So my question is: For the model to work do you need to preprocess input text based on the relarion you want to extract? So if I want to extract the relation between PERSON and GPE/LOC do I only need to feed sentences that contains a PERSON and a GPE/LOC?

And how does the model deal with sentences where entities co-occure but do not have a relation?

Thanks,
Arno

**Comment by rmitsch** [comment]
Hi @AGudden!

> ...for some sentences I got following error in my terminal: IndexError: tuple index out of range
Closer inspection showed that this is properly caused by sentences that have only 1 entity. I also noticed that the model extraxt relations between entities that are not in my example.json.file.

Could you post your stack trace (and ideally the data you're using)?

> So my question is: For the model to work do you need to preprocess input text based on the relarion you want to extract? So if I want to extract the relation between PERSON and GPE/LOC do I only need to feed sentences that contains a PERSON and a GPE/LOC?

While this depends on the model, I think it's safe to say it's not necessary to only feed sentences with the entity types whose relations to each other you're looking for. What I would recommend, to increase accuracy, is to [provide some few-shot examples](https://github.com/explosion/spacy-llm#providing-examples-for-few-shot-prompts) for those entity types.

> And how does the model deal with sentences where entities co-occure but do not have a relation?

This depends entirely on the used LLM, `spacy-llm` has no impact on this. I recommend trying out some examples to get a feel for how well this case is handled.

**Reply by AGudden** [reply]
Hi Mitsch,
thanks for your quick respone.

I attached 4 files; 'stack trace', 'examples.json' and 'input file, run_pipeline.py'

In this example I like to extract the relation between PERSON and LOCATION. In general term a person kan work, live or visits a location. In the example file I added 3 examples to define the relation. 
The first 3 sentences of the input file are reasonable close to those in the example file. You can see that the model is performing reasonable well on those 3 sentences.

The error: "IndexError: tuple index out of range" is caused by sentence 5. When you look closer you see that sentence 5 has 1 entity. I changed sentence 5 so that it has 2 entities, and added the sentence as sentence 4 in the input file. With 2 sentences the model does not generate an error. Only the relation does not makes any sense. One person lives in another person.

So does the model only works when there are 2 entities and is the best way to prefend that the model creates relations that are not in the exampel file?
[stack_trace.txt](https://github.com/explosion/spacy-llm/files/12037323/stack_trace.txt)
[input.csv](https://github.com/explosion/spacy-llm/files/12037326/input.csv)
[run_pipeline.txt](https://github.com/explosion/spacy-llm/files/12037351/run_pipeline.txt)
[examples1.txt](https://github.com/explosion/spacy-llm/files/12037384/examples1.txt)

**Reply by rmitsch** [reply]
Thanks for providing these files. Could you share your config as well?

**Reply by AGudden** [reply]
Sure!
[config.txt](https://github.com/explosion/spacy-llm/files/12038392/config.txt)

**Reply by rmitsch** [reply]
So the sentence in question, i. e. the one causing the exception, is
> Mogen die dan ook inzetten op innovatie als ze gedwongen uitkoop van boeren niet zien zitten wilde BBB-fractievoorzitter Van der Plas weten.

Correct?

**Reply by AGudden** [reply]
Yes correct

**Reply by rmitsch** [reply]
Alright, I looked into this. This is a bug on our side. The issue is that the LLM returns incorrect information here - `'{"dep": 0, "dest": 1, "relation": "WorksIn"}'` despite there being only one entity in the prompt text. We should check for the indices to be consistent with the number of entities, but we aren't doing this in the current implementation.

We'll release a patch (v0.4.2) for this soon. Probably end of this or beginning of next week. I'll update here once this has been done.

As a note - your examples are currently annoted incorrectly:
```
Laura[ENT0:PERSON] heeft een huis gekocht in Amste[ENT1:GPE]rdam.
Michael[ENT0:PERSON] is op een stedentrip in Barcel[ENT1:GPE]ona.
Peter[ENT0:PERSON] werk momenteel in Groni[ENT1:GPE]ngen.
```
You'll want to change your `examples.json` to 
```json
{"text": "Laura heeft een huis gekocht in Amsterdam.", "ents": [{"start_char": 0, "end_char": 5, "label": "PERSON"}, {"start_char": 28, "end_char": 41, "label": "GPE"}], "relations": [{"dep": 0, "dest": 1, "relation": "LivesIn"}]}
{"text": "Michael is op een stedentrip in Barcelona.", "ents": [{"start_char": 0, "end_char": 7, "label": "PERSON"}, {"start_char": 29, "end_char": 41, "label": "GPE"}], "relations": [{"dep": 0, "dest": 1, "relation": "Visits"}]}
{"text": "Peter werk momenteel in Groningen.", "ents": [{"start_char": 0, "end_char": 5, "label": "PERSON"}, {"start_char": 20, "end_char": 43, "label": "GPE"}], "relations": [{"dep": 0, "dest": 1, "relation": "WorksIn"}]}
```

**Reply by rmitsch** [reply]
FYI: the corresponding [PR](https://github.com/explosion/spacy-llm/pull/219) has been merged.

**Reply by AGudden** [reply]
Hi Mitsch,
Thanks for your anwer. I will keep an eye on the new release. For now I can use a workaround by uploading sentence that contains 2 entities.
I will change my example.json file, thanks for that.

**Reply by rmitsch** [reply, marked as answer]
`spacy-llm==0.4.2` is out and should avoid this issue (LLM responses pointing at entity indices higher than the number of available entities are discarded).
